File: src/main/scala/com/strings/leetcode/twopoints/Problem_125_isPalindrome.scala
package com.strings.leetcode.twopoints

import util.control.Breaks._

object Problem_125_isPalindrome {


  def main(args: Array[String]): Unit = {
    val s = "./"
    println(isPalindrome(s))
    val ss = ':'
    println(ss.isLetterOrDigit)

  }

  def isPalindrome(s: String): Boolean = {
    if(s == null || "".equals(s)) return true
    val ss = s.replace(" ","")
    var left = 0
    var right = ss.length - 1
    // scala 的while 在不满足条件时，还要继续向下执行？？
    while (left < right){
      breakable {
        if (!ss.charAt(left).isLetterOrDigit) {
          left += 1
          break()
        }
      }
      breakable {
        if (!ss.charAt(right).isLetterOrDigit) {
          right -= 1
          break()
        }
      }
      if(ss.charAt(left).toLower != ss.charAt(right).toLower) return false
      left += 1
      right -= 1
    }
    true
  }

}

File: src/main/java/leetcode/string/Problem_125_isPalindrome.java
package leetcode.string;

public class Problem_125_isPalindrome {
    public static void main(String[] args) {
        String s = "./";
        System.out.println(isPalindrome(s));
    }

    public static boolean isPalindrome(String s){
        if(s == null || "".equals(s)){
            return true;
        }
        int left = 0, right = s.length() - 1;
        while (left < right){
            if(!Character.isLetterOrDigit(s.charAt(left))){
                left ++;
                continue;
            }
            if(!Character.isLetterOrDigit(s.charAt(right))){
                right--;
                continue;
            }
            if(Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}
